### PR TITLE
feat: main functionalities added

### PR DESCRIPTION
**Pull request by dilyara4949**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+func main() {
+	cmds, out := readInput()
+	res := execute(cmds, out)
+
+	for _, line := range res {
+		fmt.Println(line)
+	}
+}
```
Comment (jiharevzahar) [inline main.go]: Let's change main() function. It should call just methods of structure. In projects it usually looks like
main() { project.Run()}
Let's do something similar

Comment (dilyara4949) [reply main.go]: I will try

File: main.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+			}
+			tmp = ""
+		} else {
+			tmp += string(c)
+		}
+	}
+	if tmp != "" {
+		ans = append(ans, tmp)
+	}
+	return ans
```
Comment (jiharevzahar) [inline main.go]: Can you replace this functions with strings.Split()?

Comment (dilyara4949) [reply main.go]: I can, but then some test cases will not processed as  expected(if we will pass several whitespaces), I will need change readInput() then

File: main.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	go func(cmds <-chan Command) {
+		defer wg.Done()
+		for cmd := range cmds {
+			switch cmd.Name {
+			case "cat":
+				continue
+			case "grep":
+				grepFunc(&input, cmd.Argument)
+			case "sort":
+				sortFunc(input)
```
Comment (jiharevzahar) [inline main.go]: Replace it with sort.Strings(input)

File: main.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+func execute(cmds <-chan Command, input []string) []string {
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func(cmds <-chan Command) {
+		defer wg.Done()
+		for cmd := range cmds {
+			switch cmd.Name {
+			case "cat":
```
Comment (jiharevzahar) [inline main.go]: Replace strings such as cat, grep, sort with const

File: main.go (hunk 2 above)
Comment (jiharevzahar) [reply main.go]: Several spaces could be deleted with strings.Trim

Comment (jiharevzahar) [reply main.go]: Probably Fields() will help you

File: main.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+func stdIn() []string {
+	reader := bufio.NewReader(os.Stdin)
+	input, err := reader.ReadString('\n')
+	if err != nil {
+		log.Fatal(err)
+	}
+	cmds := strings.Split(strings.TrimSpace(input), "|")
+	return cmds
```
Comment (jiharevzahar) [inline main.go]: return strings.Split(strings.TrimSpace(input), "|")

File: main.go (hunk 2 above)
Comment (dilyara4949) [reply main.go]: that's what I was implementing, thanks)

File: main.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+	return ans
+}
+
+func readFile(path string) []string {
+	dat, err := os.ReadFile(path)
+	if err != nil {
+		log.Fatal(err)
+	}
+	out := strings.Split(string(dat), "\n")
+	return out
```
Comment (jiharevzahar) [inline main.go]: return strings.Split(string(dat), "\n")

File: main.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+func sortFunc(input []string) {
+	sort.Strings(input)
+}
+
+func grepFunc(input *[]string, arg string) {
+	if arg == "" {
+		log.Fatal("empty arg for grep")
+	}
+	for i := len(*input) - 1; i >= 0; i-- {
```
Comment (jiharevzahar) [inline main.go]: Why loop moves backwards?

File: main.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+	var filePath string
+	out := make(chan Command, len(cmds))
+
+	go func(cmds []string, filePath *string) {
+		for i, cmd := range cmds {
+			sepCmd := Split(cmd)
+			if i == 0 && len(sepCmd) == 1 {
+				log.Fatal("filepath not given")
+			}
```
Comment (jiharevzahar) [inline main.go:32]: Put this if statement on line 28

			if len(sepCmd) == 1 {
				log.Fatal("filepath not given")
			}

File: main.go [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+				} else if name != "" {
+					if *filePath == "" && k == len(sepCmd)-1 {
+						*filePath = name
+						wg.Done()
+						continue
+					} else {
+						command.Argument = name
+					}
+				}
+			}
```
Comment (jiharevzahar) [inline main.go]: Put for loops in functions, nesting is too big in this place. Also I think you can simplify if statements

File: main.go [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	var filePath string
+	out := make(chan Command, len(cmds))
+
+	go func(cmds []string, filePath *string) {
+		for i, cmd := range cmds {
+			sepCmd := Split(cmd)
+			if i == 0 && len(sepCmd) == 1 {
```
Comment (jiharevzahar) [inline main.go]: i==0 is not obvious for other developers what will happen here. As I understand you check is filepath present here. You can replace it with `if cmd == cat && len(sepCmd) == 1 {log.Fatal}

File: main.go (hunk 7 above)
Comment (dilyara4949) [reply main.go]: to be sure, that I reached all the elements of slice

File: .idea/.gitignore
Comment (Mihail-Sika) [inline .idea/.gitignore:1]: shouldn't be included in PR, move `.idea` folder to the `.gitignorefile` at the root level

File: .idea/modules.xml
Comment (Mihail-Sika) [inline .idea/modules.xml:1]: shouldn't be included in PR, move `.idea` folder to the `.gitignorefile` at the root level

File: .idea/unix-pipeline.iml
Comment (Mihail-Sika) [inline .idea/unix-pipeline.iml:1]: shouldn't be included in PR, move `.idea` folder to the `.gitignorefile` at the root level

File: .idea/vcs.xml
Comment (Mihail-Sika) [inline .idea/vcs.xml:1]: shouldn't be included in PR, move `.idea` folder to the `.gitignorefile` at the root level

File: main.go [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+
+const (
+	Cat  = "cat"
+	Grep = "grep"
+	Sort = "sort"
+)
+
+func init() {
+	app = NewProject()
+}
```
Comment (Mihail-Sika) [inline main.go]: do not use `init` func, it's not a best practice

File: main.go [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+type Command struct {
+	Name     string
+	Argument string
+}
+
+const (
+	Cat  = "cat"
+	Grep = "grep"
+	Sort = "sort"
+)
```
Comment (Mihail-Sika) [inline main.go]: Should have own type e.g.

```go
type TxnStatus string

const (
  Open      TxnStatus = "TXN_OPEN"
  Discarded TxnStatus = "TXN_DISCARDED"
)
```

In this way we are making sure that only pre-defined values could be used

File: main.go [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+	"sync"
+)
+
+type project struct{}
+
+type Project interface {
+	ReadInput() (<-chan Command, []string)
+	Execute(cmds <-chan Command, input []string) []string
+	PrintResult(res []string)
+}
```
Comment (Mihail-Sika) [inline main.go]: Are there any reason to use an interface here?

File: main.go [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+
+func NewProject() Project {
+	return &project{}
+}
+
+func Run(p Project) {
+	cmds, out := p.ReadInput()
+	res := p.Execute(cmds, out)
+	p.PrintResult(res)
+}
```
Comment (Mihail-Sika) [inline main.go]: What the purpose of this func? It can be replaced by `main` func

File: main.go (hunk 9 above)
Comment (Mihail-Sika) [reply main.go:84]: Agreed

File: main.go [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+	for _, line := range res {
+		fmt.Println(line)
+	}
+}
+
+func (p *project) ReadInput() (<-chan Command, []string) {
+	cmds := stdIn()
+
+	var wg sync.WaitGroup
+	wg.Add(1)
```
Comment (Mihail-Sika) [inline main.go]: Is there a reason to use `WaitGroup`?

File: main.go [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+			if *filePath == "" && (len(sepCmd) == 1 || len(sepCmd) == 2 && sepCmd[0] == Grep) {
+				log.Fatal("filepath not given")
+			} else if len(sepCmd) == 0 {
+				log.Fatal("input is not correct")
+			}
+
+			command := Command{}
+			command.Name = strings.ToLower(sepCmd[0])
+			if *filePath == "" {
+				*filePath = sepCmd[len(sepCmd)-1]
```
Comment (Mihail-Sika) [inline main.go]: Is there a reason to use pointer? You can simplify parsing logic

File: main.go [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+				wg.Done()
+			}
+			if command.Name == Grep {
+				if len(sepCmd) >= 2 {
+					command.Argument = sepCmd[1]
+				} else {
+					log.Fatal("input is not correct")
+				}
+			}
```
Comment (Mihail-Sika) [inline main.go]: You can simplify parsing logic

File: main.go [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+	out := make(chan Command, len(cmds))
+
+	go func(cmds []string, filePath *string) {
+		for _, cmd := range cmds {
+			sepCmd := strings.Fields(cmd)
+			if *filePath == "" && (len(sepCmd) == 1 || len(sepCmd) == 2 && sepCmd[0] == Grep) {
+				log.Fatal("filepath not given")
+			} else if len(sepCmd) == 0 {
+				log.Fatal("input is not correct")
+			}
```
Comment (Mihail-Sika) [inline main.go]: Do not use `else if` use early return instead. Also, here yoy can simplify the logic and remove complex if statement

File: main.go [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+		log.Fatal(err)
+	}
+	return strings.Split(strings.TrimSpace(input), "|")
+}
+
+func readFile(path string) []string {
+	dat, err := os.ReadFile(path)
+	if err != nil {
+		log.Fatal(err)
+	}
```
Comment (Mihail-Sika) [inline main.go]: better to return this error and handle in the upper level

**Review by Mihail-Sika — CHANGES_REQUESTED** [review]

File: main.go (hunk 11 above)
Comment (Mihail-Sika) [reply main.go]: https://liyafu.com/2021-07-07-the-problems-of-go-init-func/

File: main.go (hunk 15 above)
Comment (dilyara4949) [reply main.go]: without it, readFile(filePath) may run before assigning filepath, which will end up with filepath is not given, even if we passed it

Comment (Mihail-Sika) [reply main.go]: in this case there is no sense for using gorutine here

File: main.go [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+			command := Command{}
+			command.Name = strings.ToLower(sepCmd[0])
+
+			if command.Name == string(Grep) {
+				command.Argument = sepCmd[1]
+			}
+
+			if filePath == "" {
+				filePath = sepCmd[len(sepCmd)-1]
+				wg.Done()
```
Comment (Mihail-Sika) [inline main.go]: better to use `defer wg.Done()`

File: main.go [hunk 21]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+func (p *project) Execute(cmds <-chan Command, input []string) []string {
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func(cmds <-chan Command) {
+		defer wg.Done()
+		for cmd := range cmds {
+			switch cmd.Name {
+			case string(Cat):
```
Comment (Mihail-Sika) [inline main.go]: you need to change type of the `Name` filed for avoiding type casting

**Review by Mihail-Sika — CHANGES_REQUESTED** [review]

File: main.go (hunk 20 above)
Comment (dilyara4949) [reply main.go]: so using goroutine as well as waitgroup is not correct there, did i get it?

**Comment by jiharevzahar** [comment]
LGTM

File: Makefile [hunk 22]
```diff
@@ -0,0 +1,6 @@
+CMD = cmd/main.go
+
+run:
+	go run $(CMD)
+build:
+	go build $(CMD)
```
Comment (jiharevzahar) [inline Makefile:6]: Add lint command please

File: internal/unix-pipeline.go [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+	"sort"
+	"strings"
+
+	app "github.com/dilyara4949/unix-pipeline"
+)
+
+const (
+	cat            app.Operation = "cat"
+	grep           app.Operation = "grep"
+	ssort          app.Operation = "sort"
```
Comment (jiharevzahar) [inline internal/unix-pipeline.go]: If name for variable same as name for import, you can just add alias for import `s "sort"`

File: internal/unix-pipeline.go [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+func readFile(path string) ([]string, error) {
+	dat, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	return strings.Split(string(dat), "\n"), nil
+}
```
Comment (jiharevzahar) [inline internal/unix-pipeline.go:122]: I think you can put public funcs to /pkd directory and private files you can put inside of /internal

File: unix-pipeline.go [hunk 25]
```diff
@@ -0,0 +1,8 @@
+package unixpipeline
+
+type Command struct {
+	Name     Operation
+	Argument string
+}
+
+type Operation string
```
Comment (jiharevzahar) [inline unix-pipeline.go]: Maybe it have to be inside of /internal?

File: file.txt [hunk 26]
```diff
@@ -0,0 +1,4 @@
+user1 data1
+user3 data3
+user2 data2
+uer55 data2
```
Comment (jiharevzahar) [inline file.txt]: Put it inside of /assets

**Pull request merged**